doc/sphinx/proofs/writing-proofs/equality.rst

```diff
@@ -0,0 +1,1213 @@
+
+      .. coqtop:: all
+
+         Goal True.
+         Time assert (id (fact 100) = fact 100) by with_strategy -1 [id] abstract reflexivity.
+         exact I.
+         Time Defined.
+
+      On small examples this sort of behavior doesn't matter, but
+      because Coq is a super-linear performance domain in so many
+      places, unless great care is taken, tactic automation using
+      :tacn:`with_strategy` may not be robustly performant when
+      scaling the size of the input.
+
+   .. warning::
+
+      In much the same way this tactic does not play well with
+      :cmd:`Qed` and :cmd:`Defined` without using :tacn:`abstract` as
+      an intermediary, this tactic does not play well with ``coqchk``,
+      even when used with :tacn:`abstract`, due to the inability of
+      tactics to persist information about conversion hints in the
+      proof term. See `#12200
+      <https://github.com/coq/coq/issues/12200>`_ for more details.
+
+Equality and inductive sets
+---------------------------
+
+We describe in this section some special purpose tactics dealing with
+Leibniz equality and inductive sets or types.
+
+.. tacn:: decide equality
+
+   Solves goals of the form :g:`forall x y : R, {x = y} + {~ x = y}`,
+   where :g:`R` is an inductive type such that its constructors do not take
+   proofs or functions as arguments, nor objects in dependent types. It
+   also solves goals of the form :g:`{x = y} + {~ x = y}` as well as variants
+   using `\\/` instead of `+`, or that reverse the order of the disjunction.
+
+.. tacn:: compare @one_term__1 @one_term__2
+
+   Compares two :n:`@one_term`\s of an
+   inductive datatype. If :g:`G` is the current goal, it leaves the
+   sub-goals :n:`@one_term__1 = @one_term__2 -> G` and :n:`~ @one_term__1 = @one_term__2 -> G`.
+   The type of the :n:`@one_term`\s must satisfy the same restrictions as in the
+   tactic ``decide equality``.
+
+.. tacn:: simplify_eq {? @destruction_arg }
+
+   Let :n:`@term` be the proof of a statement with the conclusion :n:`@term__1 = @term__2`.
+   If :n:`@term__1` and :n:`@term__2` are structurally different (in the sense
+   described for the tactic :tacn:`discriminate`), then the tactic
+   ``simplify_eq`` behaves as :n:`discriminate @term`, otherwise it behaves as
+   :n:`injection @term`.
+
+   If the current goal has form :g:`t1 <> t2`, it behaves as
+   :n:`intro @ident; simplify_eq @ident`.
+
+   :n:`simplify_eq @natural` is equivalent to :n:`intros until @natural` then
+   :n:`simplify_eq @ident` where :n:`@ident` is the identifier for the last
+   introduced hypothesis.
+
+   If :n:`@bindings` are provided in :n:`@destruction_arg`, they are used to
+   instantiate parameters or hypotheses of :n:`@term`.
+
+.. note::
+   If some quantified hypothesis of the goal is named :n:`@ident`,
+   then :n:`simplify_eq @ident` first introduces the hypothesis in the local
+   context using :n:`intros until @ident`.
+
+.. tacn:: esimplify_eq {? @destruction_arg }
```

Comment:
Is it plan to document what `destruction_arg` is?

Reply:
I'll put it here for now.  All the other tactics that use it are in `tactics.rst` (and out of date).

Reply:
Sorry, I should have noticed it earlier but since `esimply_eq` can be seen as a variant of `simplify_eq`, it would make sense to indent it and to put the `prodn` for `destruction_arg` below `simplify_eq`.

Reply:
Now, its definition is so simple, I wonder whether it wouldn't make sense to splice `@destruction_arg` into `{| @one_term_with_bindings | @natural }`. In any case, reversing the order of the two productions (since `one_term_with_bindings` seems to be the main variant) would probably make sense.